aws/resource_aws_opsworks_chef.go

```diff
@@ -0,0 +1,335 @@
+	d.Set("engine_model", server.EngineModel)     // TODO: possibly not bother with this
+	d.Set("engine_version", server.EngineVersion) // TODO: possibly not bother with this
+	d.Set("instance_profile_arn", server.InstanceProfileArn)
+	d.Set("instance_type", server.InstanceType)
+	if server.KeyPair != nil {
+		d.Set("ssh_key_pair", server.KeyPair)
+	}
+	d.Set("preferred_backup_window", server.PreferredBackupWindow)
+	d.Set("preferred_maintenance_window", server.PreferredMaintenanceWindow)
+	d.Set("security_group_ids", server.SecurityGroupIds)
+	d.Set("endpoint", server.Endpoint)
+	d.Set("arn", server.ServerArn)
+
+	return nil
+}
+
+func resourceAwsOpsworksChefCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*AWSClient).opsworkscmconn
+
+	err := resourceAwsOpsworksChefValidate(d)
+	if err != nil {
+		return err
+	}
+
+	req := &opsworkscm.CreateServerInput{
+		AssociatePublicIpAddress:   aws.Bool(d.Get("associate_public_ip_address").(bool)),
+		BackupRetentionCount:       aws.Int64(int64(d.Get("backup_retention_count").(int))),
+		DisableAutomatedBackup:     aws.Bool(d.Get("disable_automated_backup").(bool)),
+		EngineModel:                aws.String(d.Get("engine_model").(string)),
+		EngineVersion:              aws.String(d.Get("engine_version").(string)),
+		InstanceProfileArn:         aws.String(d.Get("instance_profile_arn").(string)),
+		InstanceType:               aws.String(d.Get("instance_type").(string)),
+		PreferredBackupWindow:      aws.String(d.Get("preferred_backup_window").(string)),
+		PreferredMaintenanceWindow: aws.String(d.Get("preferred_maintenance_window").(string)),
+		SecurityGroupIds:           aws.StringSlice(d.Get("security_group_ids").([]string)), // TODO: SecurityGroupIds, set?
+		ServiceRoleArn:             aws.String(d.Get("service_role_arn").(string)),
+		SubnetIds:                  aws.StringSlice(d.Get("subnet_ids").([]string)), // TODO: set?
+	}
+
+	chefPivotalKeyAttribute := opsworkscm.EngineAttribute{
+		Name:  aws.String("CHEF_PIVOTAL_KEY"),
+		Value: aws.String(d.Get("chef_pivotal_key").(string)),
+	}
+
+	chefDeliveryAdminPassword := opsworkscm.EngineAttribute{
+		Name:  aws.String("CHEF_DELIVERY_ADMIN_PASSWORD"),
+		Value: aws.String(d.Get("chef_delivery_admin_password").(string)),
+	}
+
+	req.SetEngineAttributes([]*opsworkscm.EngineAttribute{&chefPivotalKeyAttribute, &chefDeliveryAdminPassword})
+
+	// TODO: possibly make these optional
+	// TODO: security_group_ids, optional
+	// TODO: subnet_ids, optional
+
+	if sshKeyPair, ok := d.GetOk("ssh_key_pair"); ok {
+		req.KeyPair = aws.String(sshKeyPair.(string))
+	}
+
+	log.Printf("[DEBUG] Creating OpsWorks Chef server: %s", req)
+
+	var resp *opsworkscm.CreateServerOutput
+	err = resource.Retry(20*time.Minute, func() *resource.RetryError {
+		var cerr error
+		resp, cerr = client.CreateServer(req)
+		if cerr != nil {
+			if opserr, ok := cerr.(awserr.Error); ok {
+				// TODO: does this also happen with this resource?
```

Comment:
Should I err on the side of "we can add this back later if it turns out it's a problem" or should I try to recreate the circumstances described to see if it needs to stay?